selfdrive/updated.py

```diff
@@ -1,57 +1,267 @@
+  except OSError:
+    print("couldn't get update lock; is another updated running? going back to sleep...")
+    upd_lock_fd.close()
+    return
+
+  cloudlog.info("attempting git update inside staging overlay")
+  os.chdir(OVERLAY_MERGED)
+
+  # Un-set the successful flag to prevent the finalized tree from being
+  # activated later if the update fails in some way.
+  if os.path.isfile(os.path.join(FINALIZED, ".update_succeeded")):
+    os.remove(os.path.join(FINALIZED, ".update_succeeded"))
+  os.system("sync")
+
+  r = run(NICE_LOW_PRIORITY + ["git", "fetch"], OVERLAY_MERGED)
+  cloudlog.info("git fetch success: %s", r)
+
+  cur_hash = run(["git", "rev-parse", "HEAD"], OVERLAY_MERGED).rstrip()
+  upstream_hash = run(["git", "rev-parse", "@{u}"], OVERLAY_MERGED).rstrip()
+
+  cloudlog.info("comparing %s to %s" % (cur_hash, upstream_hash))
+  if cur_hash != upstream_hash:
+    cloudlog.info("git reset in progress")
+    r = [run(NICE_LOW_PRIORITY + ["git", "reset", "--hard", "@{u}"], OVERLAY_MERGED),
+         # TODO: with scons, is clean still required?
+         run(NICE_LOW_PRIORITY + ["git", "clean", "-xdf"], OVERLAY_MERGED),
+         run(NICE_LOW_PRIORITY + ["git", "submodule", "init"], OVERLAY_MERGED),
+         run(NICE_LOW_PRIORITY + ["git", "submodule", "update"], OVERLAY_MERGED)]
+    cloudlog.info("git reset success: %s", '\n'.join(r))
+
+    cloudlog.info("update successful!")
+  else:
+    cloudlog.info("nothing new from git at this time")
+
+  # Make sure the update succeeded flag lands on disk LAST, only when the
+  # local git repo and OP install state are in a consistent state.
+  os.system("sync")
+  Path(os.path.join(OVERLAY_MERGED, ".update_succeeded")).touch()
+  os.system("sync")
+
+  fcntl.flock(upd_lock_fd, fcntl.LOCK_UN)
+  upd_lock_fd.close()
+
+def update_params(with_date=False):
+  params = Params()
+  if os.path.isfile(os.path.join(FINALIZED, ".update_succeeded")):
     try:
-      r = subprocess.check_output(["git", "--no-pager", "show", "@{u}:RELEASES.md"])
+      # TODO: This shouldn't be binary
+      with open(os.path.join(FINALIZED, "RELEASES.md"), "rb") as f:
+        r = f.read()
       r = r[:r.find(b'\n\n')] # Slice latest release notes
       params.put("ReleaseNotes", r + b"\n")
-    except:
+    except Exception:
       params.put("ReleaseNotes", "")
+    params.put("UpdateAvailable", "1")
+  else:
+    params.put("UpdateAvailable", "0")
 
+  if with_date:
     t = datetime.datetime.now().isoformat()
     params.put("LastUpdateTime", t.encode('utf8'))
 
-    time.sleep(60*60)
+# **** main loop ****
+
+def main(gctx=None):
+  if not os.geteuid() == 0:
+    cloudlog.error("updated must be launched as root!")
+    exit(1)
+
+  ov_lock_fd = open(OVERLAY_LOCK, 'r')
+  try:
+    fcntl.flock(ov_lock_fd, fcntl.LOCK_EX | fcntl.LOCK_NB)
+    init_ovfs()
+  except IOError:
+    print("couldn't get overlay lock; is another updated running? skipping init and continuing in update-only mode...")
```

Comment:
Should we continue here? Maybe it's better to not run updater at all.

Reply:
At some point (I can't find it now) a Comma employee on Discord recommended someone run updated manually to force an immediate update attempt. The idea here (which is not yet completed) was to preserve that ability by running in a one-shot mode if it can't get the lock. Doing it safely will require wrapping attempt_update() in a separate lock. I can certainly throw that overboard if you want to make a design statement that updated should not be invoked outside of manager.

Reply:
We can throw that overboard if we write a logfile somewhere when the update fails. As long as we have an easy way to inspect the last updater failure it's fine.

Reply:
Okay, I will reject all attempts to start updated if the overlay lock is already held, and log something accordingly.